File: code/functions/sqrt_n.R
# Set output to match scheme and python
options(digits = 16)

# 2 global functions
average <- function(x, y) (x + y)/2

square <- function(x) x*x

# Block structured function with lexical scoping
sqrt_n <- function(x) {
    # Block function to calculate square roots by
    # Newton's method.

    good_enough <- function(guess) {
        abs(square(guess) - x) < 0.001
    }

    improve <- function(guess) {
        average(guess, x/guess)
    }
    # Recursive function
    sqrt_iter <- function(guess) {
        ifelse(good_enough(guess),
               guess,
               sqrt_iter(improve(guess)))
        }
        
    sqrt_iter(1.0)
}

File: code/functions/sqrt_n.py
# sqrt_n.py


# 2 global functions
def square(x):
    return x*x


def average(x, y):
    return (x + y)/2


# Block structured function with lexical scoping
def sqrt_n(x):
    """
    Block function to calculate square roots by
    Newton's method.
    """

    def good_enough(guess):
        return abs(square(guess) - x) < 0.001

    def improve(guess):
        return average(guess, x/guess)

    # Recursive function
    def sqrt_iter(guess):
        if good_enough(guess):
            return guess
        else:
            return sqrt_iter(improve(guess))

    return sqrt_iter(1.0)
